Source/DataSources/EntityCollection.js

```diff
@@ -30,15 +30,29 @@ define([
     };
 
     function fireChangedEvent(collection) {
+        if (collection._firing) {
+            collection._refire = true;
+            return;
+        }
+
         if (collection._suspendCount === 0) {
             var added = collection._addedEntities;
             var removed = collection._removedEntities;
             var changed = collection._changedEntities;
             if (changed.length !== 0 || added.length !== 0 || removed.length !== 0) {
-                collection._collectionChanged.raiseEvent(collection, added.values, removed.values, changed.values);
-                added.removeAll();
-                removed.removeAll();
-                changed.removeAll();
+                collection._firing = true;
+                do {
+                    collection._refire = false;
+                    var addedArray = collection._addedEntities.values.slice(0);
```

Comment:
these should use the local variables already present above, `added`, `removed`, etc.